server/src/main/java/org/opensearch/index/shard/RemoteStoreRefreshListener.java

```diff
@@ -88,46 +94,65 @@ public void afterRefresh(boolean didRefresh) {
                         this.remoteDirectory.init();
                     }
                     try {
-                        String lastCommittedLocalSegmentFileName = SegmentInfos.getLastCommitSegmentsFileName(storeDirectory);
-                        if (!remoteDirectory.containsFile(
-                            lastCommittedLocalSegmentFileName,
-                            getChecksumOfLocalFile(lastCommittedLocalSegmentFileName)
-                        )) {
+                        // if a new segments_N file is present in local that is not uploaded to remote store yet, it
+                        // is considered as a first refresh post commit. A cleanup of stale commit files is triggered.
+                        // Ideally, we want this to be done in async flow. (GitHub issue #4315)
+                        if (isRefreshAfterCommit()) {
                             deleteStaleCommits();
                         }
+                        String segmentInfoSnapshotFilename = null;
                         try (GatedCloseable<SegmentInfos> segmentInfosGatedCloseable = indexShard.getSegmentInfosSnapshot()) {
                             SegmentInfos segmentInfos = segmentInfosGatedCloseable.get();
-                            Collection<String> refreshedLocalFiles = segmentInfos.files(true);
 
-                            List<String> segmentInfosFiles = refreshedLocalFiles.stream()
+                            Collection<String> localSegmentsPostRefresh = segmentInfos.files(true);
```

Comment:
Renamed the variable name from `refreshedLocalFiles` to `localSegmentsPostRefresh` to make it more clearer but it would mean the diff is shown each time the variable is used.